query/src/cmd/compaction.rs

```diff
@@ -0,0 +1,255 @@
+    expirations: DelayQueue<u64>,
+}
+
+impl CompactionQueue {
+    fn new(timeout: Duration) -> Self {
+        Self {
+            timeout,
+            entries: HashMap::new(),
+            expirations: DelayQueue::new(),
+        }
+    }
+
+    fn insert(&mut self, table_id: u64, table_info: TableInfo) {
+        if let Some((_, key)) = self.entries.get(&table_id) {
+            self.expirations.reset(key, self.timeout);
+            self.entries.insert(table_id, (table_info, *key));
+        } else {
+            let key = self.expirations.insert(table_id, self.timeout);
+            self.entries.insert(table_id, (table_info, key));
+        }
+    }
+
+    fn poll_expired(&mut self, cx: &mut Context) -> Poll<Option<TableInfo>> {
+        match ready!(self.expirations.poll_expired(cx)) {
+            Some(key) => {
+                let table_info = self.entries.remove(key.get_ref()).unwrap().0;
+                Poll::Ready(Some(table_info))
+            }
+            None => Poll::Pending,
+        }
+    }
+
+    fn create_channel(self, ctx: Arc<QueryContext>) -> (DelayQueueInsert, DelayQueueRead) {
+        let (queue_add, mut queue_add_rx) = channel::<(u64, TableInfo)>(1);
+        let (queue_read_tx, queue_read) = channel::<Option<TableInfo>>(1);
+
+        ctx.get_storage_runtime().spawn(async move {
+            let mut task = self;
+            let mut queue_add_eof = false;
+            loop {
+                tokio::select! {
+                    // delayed item from the DelayQueue becomes available.
+                    delayed_item = task.next() => {
+                        match delayed_item {
+                            Some(item) => {
+                                // forward it to the queue_read channel.
+                                if queue_read_tx.send(Some(item)).await.is_err() {
+                                    // queue_read channel receiver side was dropped.
+                                    break;
+                                }
+                            },
+                            None => {
+                                // if the queue_add channel is gone, we're done.
+                                if queue_add_eof {
+                                    break;
+                                }
+                            },
+                        }
+                    }
+                    // an item is received from the queue_add channel.
+                    new_item = queue_add_rx.recv(), if !queue_add_eof => {
+                        match new_item {
+                            Some(item) => {
+                                // insert it into the DelayQueue.
+                                task.insert(item.0, item.1);
+                            },
+                            None => {
+                                // the queue_add channel was dropped.
+                                queue_add_eof = true;
+                            },
```

Comment:
I must have missed something,  it seems that `queue_read_tx.recv` happens just after the `queue_read_tx.send`. 

can this be collapsed into :

~~~
tokio::select! {
                      .....
                    // delayed item from the DelayQueue becomes available.
                    delayed_item = task.next() => {
                        match delayed_item {
                            Some(item) => {
                                 task.insert(item.0, item.1);
                            },
                              ...
                        }
                     }
 }                 
~~~~